### ISSUE TITLE
Add 'no op' execution mode

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/nd4j#1941]
The idea: we execute a sequence of ND4J operations without any operations actually occurring. Furthermore, we have access to information on what specifically occurs between the start and the end of the no-operation mode.

Use cases:
1. Estimating memory use without actually allocating any memory (inc. whether that array allocation occurs within a workspace or not)
2. Knowing which operations specifically are executed

Some possible complications: any operations that returned a non-INDArray result. For example, entire array reductions, get double etc; we may simply be able to return zero for these operations, however.

**Comment by raver119** [comment]
yep, we'll return stubs everywhere, since in our codebase we'll have guarantee that result of no-op run won't be used anywhere (as in, model won't be saved, etc)
